[SPARK-9057] [STREAMING] Twitter example joining to static RDD of word sentiment values

Agent007 opened the pull request:
Example of joining a static RDD of word sentiments to a streaming RDD of Tweets in order to demo the usage of the transform() method.

Commit:
[SPARK-9057] [STREAMING] Example of joining a static RDD of word sentiments to a streaming RDD of Tweets in order to demo the usage of the transform() method.

AmplabJenkins commented:
Can one of the admins verify this patch?

rxin commented:
Thanks - can you give this pull request a more descriptive title?

Agent007 commented:
Done. Thanks for the feedback @rxin Let me know if there's anything else that you need.

rxin commented:
cc @tdas 

Would this be more prominent & useful as an example, or more useful as a blog post?

Commit:
[SPARK-9057] [STREAMING] Java Twitter Example of joining a static RDD of word sentiments to a streaming RDD of Tweets in order to demo the usage of the transform() method.

brennonyork commented:
It looks like [SPARK-9057](https://issues.apache.org/jira/browse/SPARK-9057) requires an example of `DStream.transform` done in all three primary Spark langs (java, scala, and python). This is a great start, but will need a Python and Scala example as well to be considered complete.

brennonyork commented:
Is there really a need to commit the `twitter_sentiment_list.txt` file? I would think we could grab it remotely on execution rather than commit it into the source code base.

Agent007 commented:
Thanks for the feedback @brennonyork The Scala example was done in a previous commit on this branch. When I tried to do the same thing in Python, I got the error message: "It appears that you are attempting to broadcast an RDD or reference an RDD from an action or transformation. RDD transformations and actions can only be invoked by the driver, not inside of other transformations; for example, rdd1.map(lambda x: rdd2.values.count() * x) is invalid because the values transformation and count action cannot be performed inside of the rdd1.map transformation. For more information, see SPARK-5063."

Commit:
[SPARK-9057] [STREAMING] Python example of joining a static RDD of word sentiments to a streaming RDD of words in order to demo the usage of the transform() method. Because there currently isn't a Python equivalent of TwitterUtils, I opted to use the streaming network_wordcount.py example for inspiration.

Commit:
[SPARK-9057] [STREAMING] As brennanyork suggested, the word-sentiment list file is downloaded during execution of the example streaming programs instead of relying on a file committed to the repository. Other miscellaneous changes include: using the AFINN word list, correcting the sort order of displaying the most positive words, correcting the variable name from happiest60 to happiest10, and code formatting

Agent007 commented:
As @brennonyork suggested, the example programs now download the word-sentiment list file and `twitter_sentiment_list.txt` has been removed from the repo. The examples in all 3 languages are all in this branch. Is there anything else you gentlement need, @rxin & @tdas ?

Agent007 commented:
Now that 1.5 has been released, would this be ok to merge? @rxin @tdas @brennonyork Congrats on the new release, by the way!

Agent007 commented:
@tdas & @rxin - It was good meeting you gentlemen in person at the Meetup in Uber HQ. Thanks again for taking time to talk to me and review my code.

rxin commented:
cc @tdas

tdas commented on examples/src/main/java/org/apache/spark/examples/streaming/JavaTwitterTagSentiments.java:
```diff
@@ -0,0 +1,178 @@
+ * the AFINN word list (http://neuro.imm.dtu.dk/wiki/AFINN)
+ */
+public class JavaTwitterTagSentiments {
+    private JavaTwitterTagSentiments() {
```
Is this constructor necessary?

Agent007 replied:
I followed the example in https://github.com/apache/spark/blob/master/examples/src/main/java/org/apache/spark/examples/streaming/JavaKafkaWordCount.java but I can see how that may be distracting and superfluous for the purposes of showing the example. I now notice that the other Java streaming examples don't have it either. Give me a few minutes to make a new commit to remove it. Thanks for your feedback @tdas

tdas commented on examples/src/main/java/org/apache/spark/examples/streaming/JavaTwitterTagSentiments.java:
```diff
@@ -0,0 +1,178 @@
+ * Displays the most positive hash tags by joining the streaming Twitter data with a static RDD of
+ * the AFINN word list (http://neuro.imm.dtu.dk/wiki/AFINN)
+ */
+public class JavaTwitterTagSentiments {
```
This whole file should be indented with 2 spaces, to keep it consistent with other Java examples.

tdas commented on examples/src/main/java/org/apache/spark/examples/streaming/JavaTwitterTagSentiments.java:
```diff
@@ -0,0 +1,178 @@
+        System.setProperty("twitter4j.oauth.consumerSecret", consumerSecret);
+        System.setProperty("twitter4j.oauth.accessToken", accessToken);
+        System.setProperty("twitter4j.oauth.accessTokenSecret", accessTokenSecret);
+
```
Please add more comments throughout the code to explain stuff.

Commit:
[SPARK-9057] [STREAMING] Removed superfluous private constructor per feedback from tdas

tdas commented:
Jenkins, ok to test.

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #42873 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/42873/consoleFull) for   PR 8431 at commit [`fd498d4`](https://github.com/apache/spark/commit/fd498d41c83d1c065853d3dafe0c1026fba46279).

tdas commented:
Some high level comments. 
- I want to make it obvious that this is an example for joins on streams. So it would be good to have "join" in the name of the example files.
- We usually do not refer to external files (the AFN...) as we dont want our examples to silently break based on externat. How big is the file? If its few KBs then we can have it in our source code.

SparkQA commented:
[Test build #42873 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/42873/console) for   PR 8431 at commit [`fd498d4`](https://github.com/apache/spark/commit/fd498d41c83d1c065853d3dafe0c1026fba46279).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `public class JavaTwitterTagSentiments `

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/42873/
Test PASSed.

Agent007 commented:
@tdas The AFINN word list is 27.4 KB (see https://github.com/fnielsen/afinn/blob/master/afinn/data/AFINN-111.txt) Should I add the text file? I'll incorporate the rest of your feedback in the next commit.

Commit:
[SPARK-9057] [STREAMING]

Based on feedback from tdas, the relevant example class names now include the word "join" in order to be clearer that the example shows how to join a streaming RDD; the Java example is reformatted to use 2 spaces for indentation; and, more comments are included.

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

tdas commented:
Aah, so consider adding this to the spark/data/ directory. You will find
other data necessary for mllib tests in that.


On Tue, Sep 22, 2015 at 7:18 PM, UCB AMPLab <notifications@github.com>
wrote:

> Merged build started.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/apache/spark/pull/8431#issuecomment-142472821>.
>

SparkQA commented:
[Test build #42877 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/42877/consoleFull) for   PR 8431 at commit [`fbcde0c`](https://github.com/apache/spark/commit/fbcde0ca793f113ab39f9695f1c22a2cf7f7e62e).

SparkQA commented:
[Test build #42877 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/42877/console) for   PR 8431 at commit [`fbcde0c`](https://github.com/apache/spark/commit/fbcde0ca793f113ab39f9695f1c22a2cf7f7e62e).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `public class JavaTwitterHashTagJoinSentiments `

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/42877/
Test PASSed.

Commit:
[SPARK-9057] [STREAMING]

The word sentiment file for the streaming join examples is included in data/streaming as suggested by tdas

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #42909 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/42909/consoleFull) for   PR 8431 at commit [`9998df6`](https://github.com/apache/spark/commit/9998df67b1c15c7cc350ef59c10af8275b947e02).

SparkQA commented:
[Test build #42909 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/42909/console) for   PR 8431 at commit [`9998df6`](https://github.com/apache/spark/commit/9998df67b1c15c7cc350ef59c10af8275b947e02).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `public class JavaTwitterHashTagJoinSentiments `

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/42909/
Test PASSed.

Agent007 commented:
@tdas The file is now in data/streaming/ and the rest of your feedback has been incorporated. Let me know if there's anything else you need. Thanks again!

Benedicto commented on examples/src/main/scala/org/apache/spark/examples/streaming/TwitterHashTagJoinSentiments.scala:
```diff
@@ -0,0 +1,96 @@
+
+    // Print hash tags with the most positive sentiment values
+    happiest60.foreachRDD(rdd => {
+      val topList = rdd.take(10)
```
10 on this line should be 60

srowen commented:
@tdas this is an old one but looks good to go and finishes out your list of starter tasks for streaming.

srowen commented:
@Agent007 does this require adding this large data file to the repo?

Agent007 commented:
Hi @srowen Yes, it has been discussed in this pull request. The AFINN word sentiment file is only 27.4 KB and is used to calculate the Twitter hash tag sentiments for the streaming example. At first, the file was downloaded at runtime as suggested by @brennonyork but @tdas wanted the file in the spark/data directory instead.

SparkQA commented:
[Test build #2214 has started](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/2214/consoleFull) for   PR 8431 at commit [`9998df6`](https://github.com/apache/spark/commit/9998df67b1c15c7cc350ef59c10af8275b947e02).

SparkQA commented:
[Test build #2214 has finished](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/2214/console) for   PR 8431 at commit [`9998df6`](https://github.com/apache/spark/commit/9998df67b1c15c7cc350ef59c10af8275b947e02).
 * This patch **passes all tests**.
 * This patch **does not merge cleanly**.
 * This patch adds the following public classes _(experimental)_:
  * `public class JavaTwitterHashTagJoinSentiments `

Agent007 commented:
@srowen Any idea why this didn't merge cleanly? I didn't see anything wrong from the Jenkins output. And, I just tried a local merge to my local copy of master and saw no merge conflicts. I've also synchronized my forked repo with the original repo, as I hadn't done that in a while. Thanks for your help.

BenFradet commented:
Hi @Agent007, have you tried rebasing against master?

srowen commented:
There is no merge conflict, so no rebase should be necessary. Let me try it again.

SparkQA commented:
**[Test build #2216 has started](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/2216/consoleFull)** for PR 8431 at commit [`9998df6`](https://github.com/apache/spark/commit/9998df67b1c15c7cc350ef59c10af8275b947e02).

3ourroom commented:
NAVER - http://www.naver.com/
--------------------------------------------

3ourroom@naver.com 님께 보내신 메일 <Re: [spark] [SPARK-9057] [STREAMING] Twitter example joining to static RDD of word sentiment values (#8431)> 이 다음과 같은 이유로 전송 실패했습니다.

--------------------------------------------

받는 사람이 회원님의 메일을 수신차단 하였습니다. 


--------------------------------------------

SparkQA commented:
**[Test build #2216 has finished](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/2216/consoleFull)** for PR 8431 at commit [`9998df6`](https://github.com/apache/spark/commit/9998df67b1c15c7cc350ef59c10af8275b947e02).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:\n  * `public class JavaTwitterHashTagJoinSentiments `\n

srowen commented:
@tdas I'd like to merge since the example looks good, you proposed it and looks like you substantially OK'ed it above, tests pass.

srowen commented:
Merged to master
